src/Symfony/Component/Messenger/Transport/Doctrine/DoctrineReceiver.php

```diff
@@ -0,0 +1,128 @@
+use Symfony\Component\Messenger\Exception\MessageDecodingFailedException;
+use Symfony\Component\Messenger\Exception\TransportException;
+use Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface;
+use Symfony\Component\Messenger\Transport\Serialization\Serializer;
+use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;
+
+/**
+ * @author Vincent Touzet <vincent.touzet@gmail.com>
+ */
+class DoctrineReceiver implements ReceiverInterface
+{
+    private $connection;
+    private $serializer;
+    private $shouldStop = false;
+
+    public function __construct(Connection $connection, SerializerInterface $serializer = null)
+    {
+        $this->connection = $connection;
+        $this->serializer = $serializer ?? Serializer::create();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function receive(callable $handler): void
+    {
+        while (!$this->shouldStop) {
+            $this->connection->requeueMessages();
+            try {
+                $doctrineEnvelope = $this->connection->get();
+            } catch (DBALException $exception) {
+                throw new TransportException($exception->getMessage(), 0, $exception);
+            }
+
+            if (null === $doctrineEnvelope) {
+                $handler(null);
+
+                usleep($this->connection->getConfiguration()['loop_sleep'] ?? 200000);
+
+                continue;
+            }
+
+            try {
+                $envelop = $this->serializer->decode([
+                    'body' => $doctrineEnvelope['body'],
+                    'headers' => $doctrineEnvelope['headers'],
+                ]);
+            } catch (MessageDecodingFailedException $exception) {
+                $this->rejectDoctrineEnvelope($doctrineEnvelope);
+
+                throw $exception;
+            }
+
+            $envelop = $envelop->with(new DoctrineReceivedStamp($doctrineEnvelope));
+            $handler($envelop);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function ack(Envelope $envelope): void
+    {
+        $this->ackDoctrineEnvelope($this->findDoctrineEnvelope($envelope));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reject(Envelope $envelope): void
+    {
+        $this->rejectDoctrineEnvelope($this->findDoctrineEnvelope($envelope));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function stop(): void
+    {
+        $this->shouldStop = true;
+    }
+
+    private function findDoctrineEnvelope(Envelope $envelope): array
+    {
+        /** @var DoctrineReceivedStamp|null $doctrineReceivedStamp */
+        $doctrineReceivedStamp = $envelope->last(DoctrineReceivedStamp::class);
+
+        if (null === $doctrineReceivedStamp) {
+            throw new LogicException('No DoctrineReceivedStamp found on the Envelope.');
+        }
+
+        return $doctrineReceivedStamp->getDoctrineEnvelope();
+    }
+
+    private function ackDoctrineEnvelope(array $doctrineEnvelope)
```

Comment:
As you'll change the stamp to just contain the `ID`, this should just take the `string $id` as a parameter (changing the methods name).